src/attr/_make.py

```diff
@@ -1672,6 +1672,36 @@ def fields_dict(cls):
     return ordered_dict(((a.name, a) for a in attrs))
 
 
+def resolve_types(cls, globalns=None, localns=None):
+    """
+    Resolve any strings and forward annotations in type annotations.
+
+    :param type cls: Class to resolve.
+    :param globalns: Dictionary containing global variables, if needed.
+    :param localns: Dictionary containing local variables, if needed.
+
+    :raise TypeError: If *cls* is not a class.
+    :raise attr.exceptions.NotAnAttrsClassError: If *cls* is not an ``attrs``
+        class.
+    :raise NameError: If types cannot be resolved because of missing variables.
+
+    ..  versionadded:: 17.4.0
+    """
+    try:
+        # Since calling get_type_hints is expensive we cache whether we've
+        # done it already.
+        cls.__attrs_types_resolved__
+    except AttributeError:
+        import typing
+
+        hints = typing.get_type_hints(cls, globalns=globalns, localns=localns)
+        for field in fields(cls):
```

Comment:
Would it make sense at all to make the function recursive?

Reply:
Interesting. Do you mean adding something like:
```
if attr.has(hints[field.name]):
    resolve_types(hints[field.name], globalns, localns)
```

It's an interesting idea but I wonder about 2 things:
* The call overhead.  If I have 20 classes that use 1 class, then I end up calling `resolve_types` 40 times.  (Granted 19 of those calls are no-ops, but in python any function call costs)
* Whether it makes sense to pass the globalns and localns to the sub type.

Reply:
Yeah I’m not sure myself. I think it’s fine to leave it for later